exceptions.py

```diff
@@ -76,24 +76,26 @@ def _handle_exception(self, main_client, **kwargs):
         # Let one greenlet through per region_client (returns True otherwise
         # blocks and eventually returns False)
         if _let_one_through(self, self.region_client):
-            if self.region_client is not None:
-                # We need to make sure that a different thread hasn't already
-                # reestablished to this region.
-                loc = self.region_client.host + ":" + self.region_client.port
-                if loc in main_client.reverse_client_cache:
-                    # We're the first in and it's our job to kill the client.
-                    # Purge it.
-                    logger.warn("Region server %s:%s refusing connections. Purging cache, sleeping, retrying.",
-                                self.region_client.host, self.region_client.port)
-                    main_client._purge_client(self.region_client)
-                    # Sleep for an arbitrary amount of time. If this returns
-                    # False then we've hit our max retry threshold. Die.
-                    key = self.region_client.host + ':' + self.region_client.port
-                    if not _dynamic_sleep(self, key):
-                        raise self
-            # Notify all the other threads to wake up because we've handled the
-            # exception for everyone!
-            _let_all_through(self, self.region_client)
+            try:
```

Comment:
No.

Only when it returns True, it does not release the semaphore. When it returns False, it has released the semaphore already